File: python/fastapi/src/template_api/core/database_metrics.py
from collections.abc import Callable
from dataclasses import dataclass

from prometheus_client import Counter, Gauge, Histogram

from template_api.contracts.database import AcquisitionOutcome, TransactionOutcome
from template_api.core.metrics import HttpMetrics


@dataclass
class DatabaseMetrics:
    owner: HttpMetrics
    connections: Gauge
    hold: Histogram
    sessions: Gauge
    acquisition: Histogram
    timeouts: Counter
    transactions: Counter
    transaction_duration: Histogram

    def record(self, action: Callable[[], None]) -> None:
        try:
            action()
        except Exception:
            self.owner.failed = True

    def record_acquisition(self, outcome: AcquisitionOutcome, seconds: float) -> None:
        self.record(
            lambda: self.acquisition.labels("primary", outcome.value).observe(seconds)
        )
        if outcome is AcquisitionOutcome.TIMEOUT:
            self.record(self.timeouts.inc)

    def record_transaction(self, outcome: TransactionOutcome, seconds: float) -> None:
        self.record(lambda: self.transactions.labels("primary", outcome.value).inc())
        self.record(
            lambda: self.transaction_duration.labels("primary", outcome.value).observe(
                seconds
            )
        )


def create_database_metrics(owner: HttpMetrics, limit: int) -> DatabaseMetrics:
    labels = {"role": "primary"}
    connections = Gauge(
        "db_pool_connections_in_use",
        "Connections currently checked out.",
        labels,
        registry=owner.registry,
    ).labels(**labels)
    hold = Histogram(
        "db_pool_connection_hold_seconds",
        "Time from checkout to return or detach.",
        labels,
        registry=owner.registry,
    ).labels(**labels)
    Gauge(
        "db_pool_connection_limit",
        "Configured pool size plus allowed overflow.",
        labels,
        registry=owner.registry,
    ).labels(**labels).set(limit)
    sessions = Gauge(
        "db_sessions_active",
        "Sessions owned by the session provider.",
        labels,
        registry=owner.registry,
    ).labels(**labels)
    acquisition = Histogram(
        "db_connection_acquire_seconds",
        "Explicit connection acquisition including connect and validation costs.",
        ("role", "outcome"),
        registry=owner.registry,
    )
    timeouts = Counter(
        "db_pool_timeouts_total",
        "Pool timeouts at explicit connection acquisition.",
        labels,
        registry=owner.registry,
    ).labels(**labels)
    transactions = Counter(
        "db_transactions_total",
        "Business transaction outcomes after finalization.",
        ("role", "outcome"),
        registry=owner.registry,
    )
    duration = Histogram(
        "db_transaction_duration_seconds",
        "Business transaction duration including commit or rollback.",
        ("role", "outcome"),
        registry=owner.registry,
    )
    return DatabaseMetrics(
        owner,
        connections,
        hold,
        sessions,
        acquisition,
        timeouts,
        transactions,
        duration,
    )

File: ts/nestjs/src/observability/database.metrics.ts
import { Inject, Injectable } from '@nestjs/common';
import { Counter, Gauge, Histogram } from '@prometheus-io/client';
import { Metrics, durationBuckets } from './metrics.js';
import type { TransactionOutcome } from '../contracts/observation.contract.js';

@Injectable()
export class DatabaseMetrics {
  readonly connections: Gauge<'role'>;
  readonly limit: Gauge<'role'>;
  readonly sessions: Gauge<'role'>;
  readonly hold: Histogram<'role'>;
  readonly acquisition: Histogram<'role' | 'outcome'>;
  readonly timeouts: Counter<'role'>;
  readonly transactions: Counter<'role' | 'outcome'>;
  readonly duration: Histogram<'role' | 'outcome'>;
  constructor(@Inject(Metrics) readonly owner: Metrics) {
    const registers = [owner.registry];
    const buckets = [...durationBuckets, 10, 30, 60, 120];
    this.connections = new Gauge({
      name: 'db_pool_connections_in_use',
      help: 'Worker connections checked out.',
      labelNames: ['role'],
      registers,
    });
    this.limit = new Gauge({
      name: 'db_pool_connection_limit',
      help: 'Maximum worker connections.',
      labelNames: ['role'],
      registers,
    });
    this.sessions = new Gauge({
      name: 'db_sessions_active',
      help: 'Connection leases including acquisition wait.',
      labelNames: ['role'],
      registers,
    });
    this.hold = new Histogram({
      name: 'db_pool_connection_hold_seconds',
      help: 'Worker checkout through return.',
      labelNames: ['role'],
      registers,
      buckets,
    });
    this.acquisition = new Histogram({
      name: 'db_connection_acquire_seconds',
      help: 'Worker acquisition including creation and validation.',
      labelNames: ['role', 'outcome'],
      registers,
      buckets,
    });
    this.timeouts = new Counter({
      name: 'db_pool_timeouts_total',
      help: 'Worker pool acquisition timeouts.',
      labelNames: ['role'],
      registers,
    });
    this.timeouts.labels('primary').inc(0);
    this.transactions = new Counter({
      name: 'db_transactions_total',
      help: 'Business transactions after finalization.',
      labelNames: ['role', 'outcome'],
      registers,
    });
    this.duration = new Histogram({
      name: 'db_transaction_duration_seconds',
      help: 'Business transaction duration through commit or rollback.',
      labelNames: ['role', 'outcome'],
      registers,
      buckets,
    });
    for (const outcome of ['committed', 'rolled_back', 'failed']) {
      this.transactions.labels('primary', outcome).inc(0);
    }
  }
  transaction(outcome: TransactionOutcome, seconds: number): void {
    this.owner.record(() => {
      this.transactions.labels('primary', outcome).inc();
      this.duration.labels('primary', outcome).observe(seconds);
    });
  }
}
